File: Decorator/decorator.cpp
#include<bits/stdc++.h>
using namespace std;
class Coffee{
    public:
    virtual string getDescription();
    virtual double getCost();
};

class PlainCoffee:public Coffee{
    public:
    string getDescription()override{
        return "Plain Coffee";
    }
    double getCost override{
        return 2.0;
    }
};
class CoffeeDecorator:public Coffee{
    protected:
     Coffee* decoratedCoffee;
     public:
     CoffeeDecorator(Coffee*coffee):decoratedCoffee(coffee){}
     virtual string getDescription() const override{
        return decoratedCoffee->getDescription();
     }
     virtual double getCost()const override{
        return decoratedCoffee->getCost();
     }
};
class MilkDecorator:public CoffeeDecorator{
    public:
    MilkDecorator(Coffee*coffee):CoffeeDecorator(coffee){}
    string getDescription()const override{
        return CoffeeDecorator::getDescription()+",Milk";
    }
    double getCost()const override{
        return CoffeeDecorator::getCost()+0.5;
    }
};
class SugarDecorator:public CoffeeDecorator{
    public:
    SugarDecorator(Coffee*coffee):CoffeeDecorator(coffee){}
    string getDescription()const override{
        return CoffeeDecorator::getDescription()+", Sugar";
    }
    double getCost()const override{
        return CoffeeDecorator::getCost()+0.2;
    }
};
int main(){
    Coffee*coffee=new PlainCoffee();
    cout<<"Description "<<coffee->getDescription()<<endl;
    cout<<"Cost $"<<coffee->getCost()<<endl;

    Coffee*milkCoffee=new MilkDecorator(new PlainCoffee());
    cout<<"\nDescription"<<milkCoffee->getDescription()<<endl;
    cout<<"Cost $"<<milkCoffee->getCost()<<endl;

    Coffee*sugarMilkCoffee=new SugarDecorator(new MilkDecorator(new PlainCoffee()));
    cout<<"\n Description "<<sugarMilkCoffee->getDescription()<<endl;
    cout<<"Cost $"<<sugarMilkCoffee->getCost()<<endl;
    delete coffee;
    delete milkCoffee;
    delete sugarMilkCoffee;
    return 0;
}

File: Decorator/decorator.py
from abc import ABC, abstractmethod
class Coffee(ABC):
    @abstractmethod
    def get_description(self):
        pass
    
    @abstractmethod
    def get_cost(self):
        pass

class PlainCoffee(Coffee):
    def get_description(self):
        return "Plain Coffee"
    
    def get_cost(self):
        return 2.0
    
class CoffeeDecorator(Coffee):
    def __init__(self,decorated_coffee):
        self.decorated_coffee=decorated_coffee
    
    def get_description(self):
        return self.decorated_coffee.get_description()

    def get_cost(self):
        return self.decorated_coffee.get_cost()
    
class MilkDecorator(CoffeeDecorator):
    def get_description(self):
        return self.decorated_coffee.get_description() + " Milk"
    def get_cost(self):
        return self.decorated_coffee.get_cost()+ 0.5
    
